### PR TITLE
fix(coedit): keep a session's Yjs lineage across a reconnect

### PR DESCRIPTION
**Pull request by Bo-Onyx**
Fixes the page-duplication corruption reported in Slack (a page "copied itself MANY times over", `cost model` appearing 1200+ times).

## Scale of it

Scanning every session that carries a body, for consecutive sessions on one path where the body grew ≥1.8×:

```
14 doubling events, 3 pages, 3 days

07-31  test-images.md                    144 →   284 →   564
08-01  Individual Notes/Bo TODO.md    28,098 → 56,098 → 112,098 → 224,099
08-02  …Value Proposition & Cost Model  2,736 → 6,184 → … → 799,576  (8 in a row)
```

Nothing bounded it. The gaps shrink as it goes, because a multi-megabyte document makes the next event-loop stall likelier, which makes the next doubling likelier.

## Cause

A session's document identity **is** its Yjs lineage. `seed_doc_from_markdown` builds a bare `Doc()`, which mints a random client id, and Yjs merges by item id — `(client_id, clock)` — never by content.

`open_session` matched only `status == ACTIVE`, so a closed session was never reused. Sequence:

1. The participant lease expires (60s, scanned every minute); the session is clean with nobody attached, so it's closed. Attached sockets then fail their next auth check and drop.
2. The browser's 3s reconnect loop fires **within the same effect run**, so it still holds its `Y.Doc` (`hooks.ts:420` is outside the loop).
3. No ACTIVE row exists → a new session → seeded fresh from markdown → **new lineage**.
4. The server offers SYNC_STEP1; the client replies with every update the server appears to lack. The server has never heard of the client's lineage, so that reply is the **entire document**.
5. The server applies it with no state-vector check. Nothing dedupes — two copies of every block, in **one** update row. This is why sessions were observed with `ydoc_seq = 1` already holding a doubled document.
6. Checkpoint commits 2N; the next cycle seeds from 2N.

The frontend was written against an invariant the backend doesn't provide. `hooks.ts:225-229` says keeping the doc across reconnects "is now possible (**the server never changes document identity**)", and `hooks.ts:96-99` asserts a reconnect "lands back on the same session_id".

## Fix

**`open_session` reactivates a matching closed session** instead of inserting a new row, when it still describes the page as committed: `base_sha` equal to the caller's HEAD (any commit from any writer moves HEAD and disqualifies it), a snapshot present, and clean (`ydoc_seq == ydoc_checkpointed_seq` — a dirty session was closed with work outstanding and reviving it would resurrect unverified updates). Same lineage ⇒ the client's reply is a no-op, and text typed while disconnected now merges instead of being dropped. It also makes the frontend's existing assumption true rather than teaching the client to cope with it being false.

**`checkpoint_session` refuses to commit implausible growth** — more than double across ≤3 update rows, above a 4 KB floor. A backstop, not the fix: it only ever declines, so a false positive defers a checkpoint while a false negative corrupts a page. It would have capped this incident at one doubling.

## Testing

The exchange is reproduced at the codec level, with no websocket, no session churn and no timing:

```python
client_doc = seed_doc_from_markdown(body)
server_doc = seed_doc_from_markdown(body)      # same text, different lineage
server_doc.apply_update(client_doc.get_update(server_doc.get_state()))
assert reconstruct_body(server_doc) == body + body
```

with the shared-lineage case asserted to be a no-op beside it. That test would have failed the day the room-less backend landed. Plus the reuse rule's four branches (match, moved-on, dirty/snapshotless, other-path) and the growth guard's bounds.

Full backend suite: 2305 passed, 2 skipped. ruff + basedpyright clean.

## Not in scope

Why a heartbeat lapses past 60s with a tab open — that's what closes the session in the first place, and it's still unproven (event-loop stalls from large pycrdt work are the leading candidate, and it's self-reinforcing). It makes churn *frequent*, not *corrupting*: with the lineage preserved, a churn is just a reconnect. The client-side identity check is the remaining defense in depth.

### CODE REVIEW

File: backend/app/wiki/coedit_checkpoint.py [hunk 1]
```diff
@@ -271,6 +294,18 @@ def checkpoint_session(session_id: int) -> CheckpointOutcome | None:
...
+                    "coedit checkpoint: refusing to commit implausible growth for %s "
+                    "(session %s): %d -> %d bytes across %d update(s); leaving the "
+                    "session dirty rather than committing likely-duplicated content",
+                    path,
+                    session_id,
+                    len(base_body),
+                    len(body),
+                    replayed_seq - sess.ydoc_checkpointed_seq,
+                )
+                return None
```
[the remark this answers is not in the document]
Comment (Bo-Onyx) [reply backend/app/wiki/coedit_checkpoint.py]: Valid, and worse than described — fixed in 946c498d by replacing the heuristic rather than tuning it.

You are right that a paste can legitimately add a page-length of content to a >=4KB page in one update. And the consequence compounds: replaying the same log rebuilds the same document, so `replayed_seq - ydoc_checkpointed_seq` is identical on every later attempt and the refusal is permanent, not deferred. That also means a checkpoint attempt every minute forever, and — as you note — an explicit save reporting success while nothing reached git. A guard that can permanently strand a real edit is worse than the corruption it was watching for.

Replaced with the actual signature: two top-level blocks carrying the same `_blockId`. Verified against a real merged pair — two seeds of the same body, the client-side sync reply applied, and both blocks come back as `b0`. Editing cannot reach that state, since freshly typed blocks carry no id and take `checkpoint_body`'s `orig is None` path, so there is no legitimate edit left to refuse. It is independently known-invalid on the client for the same reason (`UniqueBlockIdentity` in `blocks.ts` — two nodes sharing an id resolve to the same original range).

Detection is now terminal instead of a retry: log at error and close the session, so the next reader seeds from the last good commit and editing recovers, rather than the page being stuck behind a checkpoint that can never succeed.

Also added the negative test this implies — a 500-block document and a one-block document both report no duplicates, so growth alone never trips it.

**Comment by Bo-Onyx** [comment]
@greptile

File: backend/app/wiki/coedit_checkpoint.py [hunk 2]
```diff
@@ -226,6 +255,28 @@ def checkpoint_session(session_id: int) -> CheckpointOutcome | None:
...
+            # reader seeds afresh from the last good commit — the page keeps
+            # that state instead of gaining a duplicate, and editing recovers.
+            # Whatever this session held past its watermark is lost, which is
+            # the intent: in this state that content *is* the duplication.
+            log.error(
+                "coedit checkpoint: %s (session %s) holds duplicate top-level block ids "
+                "%s — refusing to commit and closing the session; the page stays at its "
+                "last checkpoint. This should be unreachable: see open_session's lineage "
+                "reuse rule.",
+                path,
```
[the remark this answers is not in the document]
Comment (Bo-Onyx) [reply backend/app/wiki/coedit_checkpoint.py:312]: Valid — traced it and you're right, so `ec8d16f8` changes the response from refuse-and-close to repair.

The loop is exactly as you describe: close the dirty session → next open can't reuse it (reuse demands a clean watermark) → fresh lineage → the browser's retained document merges its copy in again → detect → close → repeat, with that session's uncheckpointed work stranded each time.

I checked whether any server-only remedy escapes it. Resetting the server to the last good snapshot does not — the duplicated content lives in the *client's* document, so it comes straight back on the next sync:

```
client doc is corrupt:        ['b0']
re-seed / reset then sync ->  ['b0']      # loops, as you said
after server-side dedupe   ->  []
client after the broadcast ->  []         # converges
```

Deleting the repeats does escape it. A Yjs delete addresses the same items the client holds, so the update `checkpoint_session` already broadcasts at the end repairs the client's own document — no reload, and nothing left to re-send on the next reconnect. Verified directly against pycrdt, and that exchange is now a test: corrupt the client, repair server-side, apply the broadcast, assert both are clean.

So the branch now drops every repeat of a top-level `_blockId` — keeping the copy that matches committed state — logs at error, and commits the repaired document through the normal path. The copies are identical when this arises from a lineage merge, so keeping the first is deterministic rather than a judgement call; if one had been edited afterwards those edits are lost, which is still better than a page that can never be saved.

Worth noting this is the second time the guard's failure mode was the problem rather than its detection: refusing to commit looked safe both times and wasn't.

**Comment by Bo-Onyx** [comment]
@greptile
